ocaml/xapi/rpm.ml

```diff
@@ -234,6 +212,61 @@ module Pkg = struct
     in
     compare_segments (normalize s1) (normalize s2)
 
+  let rec compare_version_strings s1 s2 =
+    (* Compare versions or releases of RPM packages
+     * I.E. for "libpath-utils-0.2.1-29.el7.x86_64" and
+     * "libpath-utils-0.2.1a-30.el7.x86_64",
+     * this function compares:
+     * versions between "0.2.1" and "0.2.1a", or
+     * releases between "29.el7" and "30.el7".
+     * More examples:
+     *  "1.2.3" "<" "1.2.4"
+     *  "1.2.3" "=" "1.2.3"
+     *  "1.2.3" ">" "1.2"
+     *  "1.0011" ">" "1.9"
+     *  "1.05" "=" "1.5"
+     *  "1.0" ">" "1"
+     *  "1.0" ">" "1.a"
+     *  "2.50" ">" "2.5"
+     *  "XS3" "<" "xs2"
+     *  "1.2.3" ">" "1.2.3a"
+     *  "xs4" "=" "xs.4"
+     *  "2a" "<" "2.0"
+     *  "2a" "<" "2b"
+     *  "1.0" ">" "1.xs2"
+     *  "1.0_xs" "=" "1.0.xs"
+     *  "1.xs8" ">" "1.xs8~2_1"
+     *  "1.2.3" ">" "1.2.3~beta"
+     * Some corner cases that don't follow standard RPM versioning conventions
+     * with tilde:
+     *  "1.2.3~rc1~beta" "<" "1.2.3~rc1"
+     *  "1.2.3~" "<" "1.2.3"
+     *)
+    match (Astring.String.cut ~sep:"~" s1, Astring.String.cut ~sep:"~" s2) with
+    | None, None ->
+        compare_no_tilde_version_strings s1 s2
```

Comment:
A general approach of parsing is preferred than the specific functions.

For example of "1.2.3~rc1~beta" "<" "1.2.3~rc1", the `segment` lists would be:
`[Int 1; Int 2; Int 3; Tilde; Str "rc"; Int 1; Tilde; Str "beta"]` and `[Int 1; Int 2; Int 3; Tilde; Str "rc"; Int 1]`. It would be easy to update the pattern matching in `compare_segments`.

Reply:
Accept

Reply:
I support the proposal to parse a string into tagged components and to define order over these lists.

Reply:
According to the comment, I split the version string to segment of Int, Str and Tilde and fix some errors. I split the commits to fix segment division and support tilde. Because there is a big difference from the initial version, I push -f directly for not affecting the review experience.

Reply:
I don't know if the tilde could be considered just a character in a `Str` and you would get the desired order.

Reply:
@lindig, in your example, bash command `sort` thinks `~abc` < `abc` because the locale settings.  For example in the en_US.UTF-8 locale, the tilde character is often treated as less than alphanumeric characters.
In `man sort`
> *** WARNING *** The locale specified by the environment affects sort order.  Set LC_ALL=C to get the traditional sort order that uses native byte values.

And in OCaml, String.compare uses lexicographical order.
``` 
utop # List.sort String.compare ["ab"; "abc"; "~abc"];;
- : string list = ["ab"; "abc"; "~abc"]
```

Tilde is more complicated than just a character in a `Str`.  It may be along with letters or numbers, and even used separately. You can find more examples in the ut.
Generally, tilde (and the part after it) in the version string stands for pre-release:
1. Compare the formal version part first then compare the pre-release version part.
2. Version with tilde is earlier than the same version without tilde.

Reply:
It the end, a total order needs to be defined between two arbitrary versions. Hence, we can't rely on "same versions without a tilde" because this is just a special case. So I would like to see this total order being defined over lists of `Str`, `Int` and `Tilde` or similar.

Reply:
Of course. It's just a general statement to help comprehension. The function `compare_segments`  in the code define how to compare segment list.
Function  `compare_version_segment ` define how to compare single segment.

```ocaml
  type order = LT | EQ | GT

  type version_segment = Int of int | Str of string | Tilde

  let order_of_int = function 0 -> EQ | r when r > 0 -> GT | _ -> LT

  let rec compare_segments l1 l2 =
    match (l1, l2) with
    | c1 :: t1, c2 :: t2 -> (
      match compare_version_segment c1 c2 with
      | EQ ->
          compare_segments t1 t2
      | r ->
          r
    )
    | Tilde :: _, [] ->
        LT
    | [], Tilde :: _ ->
        GT
    | _ :: _, [] ->
        GT
    | [], _ :: _ ->
        LT
    | [], [] ->
        EQ

  let compare_version_segment s1 s2 =
    match (s1, s2) with
    | Int i1, Int i2 ->
        Int.compare i1 i2 |> order_of_int
    | Str s1, Str s2 ->
        String.compare s1 s2 |> order_of_int
    | Tilde, Tilde ->
        EQ
    | Int _, Str _ ->
        GT
    | Str _, Int _ ->
        LT
    | Tilde, _ ->
        LT
    | _, Tilde ->
        GT
```
I think the code is clear to specify the order over segment lists.

Reply:
The code is clear; this needs to be tested against the examples to show it has the desired order.

Reply:
The description is just like code :)
(1) Compare two segments:
For different kinds of segments:
Int > Str > Tilde
If they are the same kind, just compare the value (string or int)

(2) Compare two segment lists:
Compare segments in the two lists in order as (1).
If the order is determined (Not Eq)  -> end.
If they are equal, compare the next segments from the lists recursively.
If one list is empty, then Str/Int > Empty > Tilde -> end